### PR TITLE
[IMP] mail: Show trigger message before automatic tracking template response, notify on response to template

### PR DESCRIPTION
**Pull request by reth-odoo**
When setting up an alias, if a tracked value was set to send a mail template on creation of a record, the template would be marked as sent before the original message was received.
This commit puts them back into order by processing the reception of the original message within the method of creation of the record.

Responding to that template with an e-mail would not notify the followers of that record.
This template is now marked as mt_comment.

Task-2834304
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_thread.py [hunk 1]
```diff
@@ -443,6 +439,24 @@ def _check_can_update_message_content(self, message):
         if not message.message_type == 'comment':
             raise exceptions.UserError(_("Only messages type comment can have their content updated"))
 
+    def _set_default_fields(self, fields):
```
Comment (MiquelRForgeFlow) [inline addons/mail/models/mail_thread.py:442]:
```suggestion
    def _set_default_fields(self, fields_list):
```
"fields" is the one from "odoo import fields"
thus, in my opinion is better to do like in default_get (https://github.com/odoo/odoo/blob/15.0/odoo/models.py#L1370)

File: addons/mail/models/mail_thread.py [hunk 2]
```diff
@@ -515,6 +529,19 @@ def _track_get_fields(self):
 
         return model_fields and set(self.fields_get(model_fields))
 
+
+    def _track_get_changes(self, fields):
```
Comment (MiquelRForgeFlow) [inline addons/mail/models/mail_thread.py:533]:
```suggestion
    def _track_get_changes(self, fields_list):
```

**Comment by thiryr** [comment]
Applied changes and squashed commits

File: addons/mail/models/mail_thread.py [hunk 3]
```diff
@@ -601,6 +606,8 @@ def _message_track_post_template(self, changes):
         for _field_name, (template, post_kwargs) in templates.items():
             if not template:
                 continue
+            # make discussion so followers see responses (not message_type comment to prevent editing)
```
Comment (std-odoo) [inline addons/mail/models/mail_thread.py:609]: shouldn't we have the right values from `_track_template` ? (here you will break all overwrite of `_track_template`, like this one)
```
'subtype_id': self.env['ir.model.data']._xmlid_to_res_id('appointment.mt_calendar_event_canceled')
```

**Review by std-odoo — COMMENTED** [review]
@thiryr Hi :) 

Sorry for the delay, I look super clean now :) I still have 2/3 comments

Thank you :)

File: odoo/tests/common.py [hunk 4]
```diff
@@ -819,6 +827,13 @@ def setUpClass(cls):
     def setUp(self):
         super(SingleTransactionCase, self).setUp()
         self.env.flush_all()
+        # reset precommits
+        precommit_funcs = collections.deque([func for func in self.env.cr.precommit._funcs])
+        precommit_data = {key: value for (key, value) in self.env.cr.precommit.data.items()}
```
Comment (std-odoo) [inline odoo/tests/common.py:832]:
```suggestion
        precommit_data = dict(self.env.cr.precommit.data.items())
```

File: odoo/tests/common.py [hunk 5]
```diff
@@ -773,6 +773,14 @@ def setUp(self):
         self.addCleanup(self.registry.clear_caches)
         self.addCleanup(self.env.clear)
 
+        # reset precommits
+        precommit_funcs = collections.deque([func for func in self.env.cr.precommit._funcs])
+        precommit_data = {key: value for (key, value) in self.env.cr.precommit.data.items()}
```
Comment (std-odoo) [inline odoo/tests/common.py:778]: same here

File: odoo/tests/common.py [hunk 6]
```diff
@@ -819,6 +827,13 @@ def setUpClass(cls):
     def setUp(self):
         super(SingleTransactionCase, self).setUp()
         self.env.flush_all()
+        # reset precommits
+        precommit_funcs = collections.deque([func for func in self.env.cr.precommit._funcs])
+        precommit_data = {key: value for (key, value) in self.env.cr.precommit.data.items()}
+        def setPrecommit(funcs, data):
```
Comment (std-odoo) [inline odoo/tests/common.py:833]: can you add a little comment to explain why it's necessary ?

**Review by std-odoo — COMMENTED** [review]
@thiryr Hi :) 

Sorry for the delay, looks good overall, I just have a few minors comments

Good work :) 

Edit: you have a conflict, you might want to rebase

File: addons/test_mail/tests/test_message_track.py [hunk 7]
```diff
@@ -130,6 +142,33 @@ def test_message_track_template_at_create(self):
...
+        custom_values = {'name': 'Test', 'customer_id': self.user_admin.partner_id.id,
+                         'mail_template': self.env.ref('test_mail.mail_test_ticket_tracking_tpl').id}
+        self.env['mail.alias'].create({
+            'alias_name': 'groups',
+            'alias_user_id': False,
+            'alias_model_id': test_model.id,
+            'alias_contact': 'everyone',
+            'alias_defaults': custom_values})
+        record = self.format_and_process(MAIL_TEMPLATE, '"Sylvie Lelitre" <test.sylvie.lelitre@agrolait.com>',
+                                         'groups@test.com', target_field='customer_id', subject=custom_values['customer_id'],
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:158]: Mmmh, is it a trick to "find back the created record" ? have to say it is a bit strange :D but if it works ...

File: addons/test_mail/tests/test_message_track.py [hunk 8]
```diff
@@ -130,6 +142,33 @@ def test_message_track_template_at_create(self):
...
+
+        # Should be trigger message and response template
+        self.assertEqual(len(record.message_ids), 2)
+        messages = list(record.message_ids)
+        messages.sort(key=lambda msg: msg.id)
+        trigger = messages[0]
+        template = messages[1]
+        self.assertIn('Please call me as soon as possible this afternoon!', trigger.body)
+        self.assertIn('Hello', template.body)
+        self.assertEqual(template.mail_ids.recipient_ids, self.user_admin.partner_id)
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:170]: Could use "assertNotified" tool to skip the low-level check on mail recipients probably ?

File: addons/test_mail/tests/test_message_track.py [hunk 9]
```diff
@@ -130,6 +142,33 @@ def test_message_track_template_at_create(self):
...
+        self.flush_tracking()
+
+        # Should be trigger message and response template
+        self.assertEqual(len(record.message_ids), 2)
+        messages = list(record.message_ids)
+        messages.sort(key=lambda msg: msg.id)
+        trigger = messages[0]
+        template = messages[1]
+        self.assertIn('Please call me as soon as possible this afternoon!', trigger.body)
+        self.assertIn('Hello', template.body)
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:169]: We could probably check the complete rendered body (as it is not that long) ? Just because "hello" is a bit short and not crytal clear as a comparison :D

File: addons/test_mail/tests/test_message_track.py [hunk 10]
```diff
@@ -84,30 +86,40 @@ def test_message_track_subtype(self):
...
     def test_message_track_template(self):
         """ Update some tracked fields linked to some template -> message with onchange """
-        self.record.write({'mail_template': self.env.ref('test_mail.mail_test_ticket_tracking_tpl').id})
-        self.assertEqual(self.record.message_ids, self.env['mail.message'])
-
-        with self.mock_mail_gateway():
-            self.record.write({
-                'name': 'Test2',
-                'customer_id': self.user_admin.partner_id.id,
+        sender_email = 'sender@test.lan'
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_message_track.py:89]: Why are we changing the test here actually ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, finally reviewed it ! Thanks for the PR, fix globally LGTM, some comments (mainly on tests :) ) .

Cheers !

File: addons/test_mail/tests/test_message_track.py (hunk 8 above)
Comment (thiryr) [reply addons/test_mail/tests/test_message_track.py:179]: We can't because templated track messages are sent as mass mails -> no notification

File: addons/test_mail/tests/test_message_track.py (hunk 10 above)
Comment (thiryr) [reply addons/test_mail/tests/test_message_track.py]: :thinking: hu... why indeed..

I guess it might have been useful when I added that new notification_type but I really should just have added a test either way. Doesn't seem to add anything as it stands so I'll just remove it (and the test_model I'd borrowed from you)

File: addons/mail/models/mail_thread.py [hunk 11]
```diff
@@ -1625,7 +1630,9 @@ def message_parse(self, message, save_original=False):
         parent_message = self._get_parent_message(msg_dict)
         if parent_message:
             msg_dict['parent_id'] = parent_message.id
-            msg_dict['is_internal'] = bool(parent_message.subtype_id and parent_message.subtype_id.internal)
+            # If responding to a bot message, make comment as nobody would be notified otherwise
+            parent_sent_by_bot = parent_message.author_id == self.env['res.users'].browse(SUPERUSER_ID).partner_id
+            msg_dict['is_internal'] = bool(parent_message.subtype_id and parent_message.subtype_id.internal and not parent_sent_by_bot)
```
[the remark this answers is not in the document]
Comment (thiryr) [reply addons/mail/models/mail_thread.py]: :thinking: I know I'm the one who proposed the heuristic but the more I look at it the less safe it looks...

Maybe we should just change all the automated templates to explicitly post discussions? It's kind of a hack for a hack right now I guess since they're notifications to avoid notifying followers

At the risk of postponing this a bit more

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py]: Yeah, actually we don't want to send automated messages in project to all followers (as you may have template in each stage). But we want to receive the answers ... 

Don't remember all discussion, but we could use a specific field for that instead of hacking the answer mechanism, indeed. But not sure what to do...

If we define a new subtype: yet another subtype people would have to manage. There are already several default subtypes (note, activity, comment). Moreover subtypes are used for finding recipients and display tweaks, not really to define the global management of the message.
Have a specific field on message for that ('reply_do_notify'), but not a fan of adding a column for that purpose. It will be ignored by most messages.
Have a new message_type. Currently we have comment, notification, email, sms, and it is used to characterize the source (I also think we should add mailing, for message created using the mailing mode; currently email is used for both outgoing and incoming messages, would like to split those two). We could have a new message_type ('comment_automated' ? 'comment_autoreply' ? ...) that would have its own specific behavior in mailgateway. As message has a specific source, answers to it could lead to specific behavior.

All in all, if we want to avoid the hack, would go for a specific message_type I think, seems less invasive and more coherent with current model of message.

**Review by tde-banana-odoo — COMMENTED** [review]
Zblip !

File: addons/test_mail/tests/test_message_track.py (hunk 8 above)
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_message_track.py:179]: assertMail then ?

Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_message_track.py:179]: Normally you have emails that are mocked, so using assertMail should be possible :)

File: addons/test_mail/tests/test_performance.py [hunk 12]
```diff
@@ -500,7 +500,7 @@ def test_mail_composer_w_template_form(self):
         test_template.write({'attachment_ids': [(5, 0)]})
 
         customer = self.env['res.partner'].browse(self.customer.ids)
-        with self.assertQueryCount(admin=37, employee=37):  # tm 27/27 / com 35/35
+        with self.assertQueryCount(admin=39, employee=39):  # tm 29/29 / com 37/37
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_performance.py:503]: Why do we have more queries ?

File: addons/test_mail/tests/test_mail_gateway.py [hunk 13]
```diff
@@ -1170,6 +1170,48 @@ def test_message_process_bounce_records_partner(self):
...
+        )
+
+        self.format_and_process(
+            MAIL_TEMPLATE, self.email_from, '',
+            subject='Reply to some notif',
+            extra=f'References: {some_notification_message.message_id}'
+        )
+        new_msg = self.test_record.message_ids[0]
+        self.assertEqual(new_msg.parent_id, some_notification_message)
+        self.assertTrue(new_msg.is_internal, "Responses to messages sent by anyone but odoobot should keep"
```
Comment (tde-banana-odoo) [inline addons/test_mail/tests/test_mail_gateway.py:1211]: :socks: when in doubt, I order asserts alphabetically based on field we access (aka is_internal, parent_id, subtype_id). Just to have a deterministic order. Just a remark while passing by, nothing fancy :D

File: addons/mail/models/mail_thread.py [hunk 14]
```diff
@@ -1636,17 +1642,26 @@ def message_parse(self, message, save_original=False):
...
         if msg_dict['is_bounce']:
             msg_dict.update(self._message_parse_extract_bounce(message, msg_dict))
         return msg_dict
 
+    def _message_parse_extract_from_parent(self, parent_message):
+        """Derive message values from the parent."""
+        if parent_message:
+            # if the message is sent by odoobot, we can consider it to not be internal
+            parent_is_internal = bool(parent_message.subtype_id and parent_message.subtype_id.internal)
+            parent_sent_by_odoobot = parent_message.author_id == self.env.ref('base.partner_root')
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py:1658]: Technically, should be in another commit so that we split the tracking ordering from the internal flag ? That way we can have two explanations, as those are two different issues actually.

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, final comments :)

File: addons/mail/models/mail_thread.py (hunk 14 above)
Comment (thiryr) [reply addons/mail/models/mail_thread.py]: Separated both changes. The second on is bigger now too (some changes for front-end consistency)

**Comment by tde-banana-odoo** [comment]
@robodoo r+
